### PR TITLE
Add Refunds support to DigitalRiver

### PR DESCRIPTION
**Pull request by przemyslawhaptas**
Jira: https://chargify.atlassian.net/jira/software/projects/PGT/boards/32?selectedIssue=PGT-1101

* Support refunds - both full and partial.
* DigitalRiver expects full amounts (e.g. 1.01 USD) not cents
* Remote tests utilize an order factory which looks for an existing order or creates a new one, which means it waits for the order's fulfillment for 60 seconds at most

Screenshot 2021-06-16 at 12 34 45

### CODE REVIEW

File: lib/active_merchant/billing/gateways/digital_river.rb [hunk 1]
```diff
@@ -69,6 +69,27 @@ def purchase(options)
         end
       end
 
+      def refund(money, _transaction_id, options)
+        currency = options[:currency] || currency(money)
+        params =
+          {
+            'order_id' => options[:order_id],
+            'currency' => currency.upcase,
+            'amount' => localized_amount(money, currency).to_f,
```
Comment (dbackowski) [inline lib/active_merchant/billing/gateways/digital_river.rb:78]: Is calling `.to_f` here is really necessary?
I don't see it in any other place in ActiveMerchant 🤔

File: lib/active_merchant/billing/gateways/digital_river.rb [hunk 2]
```diff
@@ -69,6 +69,27 @@ def purchase(options)
         end
       end
 
+      def refund(money, _transaction_id, options)
```
Comment (pohor) [inline lib/active_merchant/billing/gateways/digital_river.rb:72]: I assume the unused var comes from how the refund method is constructed in Conduit?
Could we maybe call the refund method in Conduit without the transaction_id and remove it here?

**Review by pohor — COMMENTED** [review]
Left a comment, otherwise 👏

File: lib/active_merchant/billing/gateways/digital_river.rb (hunk 1 above)
Comment (przemyslawhaptas) [reply lib/active_merchant/billing/gateways/digital_river.rb:78]: DR expects a float and `localized_amount) returns a string, so I followed Square doing `.to_i`: https://github.com/chargify/active_merchant/blob/master/lib/active_merchant/billing/gateways/square.rb#L187

I guess you're worried about float precision and all that? I think it's fine, it's the last point before sending the value to DR, really. 🤔

File: lib/active_merchant/billing/gateways/digital_river.rb (hunk 2 above)
Comment (przemyslawhaptas) [reply lib/active_merchant/billing/gateways/digital_river.rb:72]: Yeah, this signature is common for all calls to `refund` in AM and Conduit expects it as well.
In Conduit logic for refunds this this transaction_id is passed down a couple of functions which would all require overwritting in the DigitalRiver class. Given the current [WIP PR for Conduit](url) is so small I thought it'd be better not to change the shape of these signatures and leave the variable be.

File: lib/active_merchant/billing/gateways/digital_river.rb (hunk 1 above)
Comment (dbackowski) [reply lib/active_merchant/billing/gateways/digital_river.rb:78]: I've thought that probably DR gem requires that format after I had written this comment 😉

**Review by dbackowski — APPROVED** [review]
Nice job! 👍 🚀

File: lib/active_merchant/billing/gateways/digital_river.rb (hunk 2 above)
Comment (przemyslawhaptas) [reply lib/active_merchant/billing/gateways/digital_river.rb:72]: I think I didn't actually link that WIP PR, sorry, updated! 😅

**Pull request merged**